Fixed a handful of warnings with Visual C++ and added some defines

assarbad opened the pull request:
* The use of `_ui64toa` caused a warning in `lua_int64.h`, which could be fixed by replacing the call by one to `_ui64toa_s`. Also, this now handles failures to convert gracefully, although your mileage may vary and you may want a different kind of handling there. A second instance of this issue occurred in `winreg.c` and was fixed there.
* While `lua_opentablemt` is a C function and thus takes a variable number of parameters, I got a warning because it was missing the `upval` argument which was given on the definition, but not the declaration.
* I needed the `#ifndef lua_assert` in order to work around some issues when using the C preprocessor to _include_ the `winreg.c` file.
* Defining `LUA_REG_NO_DLL` now allows to get rid of the `DllMain` and the `__declspec(dllexport)` on the exported function.
* Furthermore defining `LUA_REG_NO_HIVEOPS` allows to suppress the Lua functions which can be used for hive operations such as loading and unloading a registry key.
* One instance of using `stricmp` instead of `_stricmp` was also fixed.

The patches are pretty straightforward, so feel free to pull them into master or don't. I just want to offer them this way.

Thanks for your module!

Best regards,

Oliver

assarbad committed:
Fixed a handful of warnings and added LUA_REG_NO_DLL define for embedding the library without DLL

assarbad committed:
Masking the hidden function names in case of LUA_REG_NO_HIVEOPS defined also from declarations

assarbad commented:
@moteus need I be concerned about AppVeyor? The build was indicated as failing even before my changes.

moteus commented:
With appveyour there problem with compile flags.
There no easy way from rockspec detect target os.
https://github.com/luarocks/luarocks/issues/655

May be you have so ideas?

assarbad commented:
Nice, I'll look more closely into this and get back to you.

http://predef.sf.net/ could be of help, among other pieces of data.

assarbad commented:
So it's true that MinGW64 (that's the one I tried) doesn't  declare `WINVER`, but that'd be natural in fact. Since it's not expected to be preset, but rather be set by the developer to target a particular Windows version:

```
>gcc -dM -E - < NUL|findstr /i win
#define _WIN32 1
#define _WIN64 1
#define __WINT_MAX__ 65535
#define __WINT_MIN__ 0
#define __WIN32 1
#define __WIN64 1
#define __WINNT 1
#define __WINNT__ 1
#define __WIN32__ 1
#define __SIZEOF_WINT_T__ 2
#define WIN32 1
#define WIN64 1
#define __WINT_TYPE__ short unsigned int
#define WINNT 1
#define __WIN64__ 1
```

But trying to reproduce the exact issue, I end up with another error in `luawin_dllerror.c`.

assarbad committed:
Trying to address the Appveyor issues with lua-winreg
- We now define KEY_WOW64_32KEY and KEY_WOW64_64KEY locally, if on MinGW
- We use variadic macros on VS2005 and newer as well as on MinGW instead of the NOP_FUNCTION construct

assarbad commented:
The issue with:

```
lua test_5_1_14.lua
lua: test_5_1_14.lua:5: bad argument #2 to 'openkey' (invalid option)
stack traceback:
	[C]: in function 'openkey'
	test_5_1_14.lua:5: in main chunk
	[C]: ?
Command exited with code 1
```

was kind of obvious, though. It arose from the fact that `KEY_WOW64_32KEY` and `KEY_WOW64_64KEY` were not always defined.

So that part was easy to fix. But I now see there's still a warning regarding `_ui64toa_s` with MinGW. So my tests with MinGW64 don't seem to be representative of whatever is running on Appveyor ...

assarbad committed:
stdlib.h is required on MinGW and should have been included even on MSVC ...

assarbad commented:
Hmm, strange. So while including `stdlib.h` _is_ the right thing, it appears as if it makes no difference with respect to `_ui64toa_s`. I wonder which exact version of MinGW is running on Appveyor. The only way to find out would be to stringify some of the defines we get from the compiler and place them as Lua variables ("constants") into the `winreg` module. This way we could output them during the tests and be sure what we're dealing with. As it stands right now, it's near impossible to figure out where `_ui64toa_s` should be coming from.

assarbad commented:
@moteus Please do not yet merge. If I understand it correctly, we need to check for `MINGW_HAS_SECURE_API` being defined in order to use `_ui64toa_s`. Otherwise we have to revert to using the insecure `_ui64toa` function. It affects the two locations pointed out in my initial post.

**NB::** It appears to be insufficient to _define_ `MINGW_HAS_SECURE_API` on the command line or so, because then it uses just a stub that doesn't implement any functionality.

assarbad committed:
Addressing the warning about _ui64toa_s on MinGW

assarbad commented:
Awesome, that did the trick. Feel free to merge this pull request. It should also fix your Appveyor issue.

assarbad commented:
Hey @moteus something you'd like me to fix/add/remove in the pull request?

zhaozg commented:
+1 for merge

And with this PR, winreg will support ming64 and cross compile
